README.md
# Gone.Bike
OpenSource, OpenData stolen bikes database, search engine and reporting tool.

Data is publicly accessible at [gone.bike/exports/](https://gone.bike/exports/)

## Project Goals
Create a global database of stolen bikes that can easily be accessed and searched,
in order to provide a fast, efficient and effective way to identify suspicious bikes
(from second hand markets, online ads, etc), using image search features and dynamic filters for a quick match.

## Features
- Receive direct reports via a detailed form (done)
- Collect and index reports from various public, local websites (WIP)
- Image search (WIP)
- Filtering search (WIP)
- Threaded discussions (TODO)
- Theft Heatmap (ROADMAP)

### Use cases
- You want to buy a second hand bike (flea market, online): you can search the db to check if the bike has been reported.
- You spot a suspicious bike and you match it in the datbase: you can report it
- Police force find stolen bikes and they can use the db to identify the owner

### Motivations
Whether you're a seasoned cyclist, a daily traveler, or an aspiring rider, you should know this feeling about bikes: they are not just things,
they are objects, yes, but we create a relationship with them. They carry us, they share sweat and tears with us and they are loyal companions that contribute to make the world a better place, reducing both noise and pollution, improving our health and mood.

Having your bike stolen is not just a monetary blow, but a personal hit to a "member" of your life, almost a friend.
The feeling of powerlessness afterwards, knowing how little you can do to recover it, thinking that your wheeled buddy is somewhere out there is frustrating.

Supported by technology, the frustation can be funneled into a good activity, beign a vigilant watcher, on the lookout to fight thieves back.



## Tech Stack
- [Astro](https://astro.build) / [VueJS](https://vuejs.org/) (Frontend)
- [Celery](https://docs.celeryq.dev/) (Backend)
- [Directus](https://directus.io) (Middleware / API)
- [Docker](https://www.docker.com/) (DevOp)
- [PostgreSQL](https://www.postgresql.org) (Backend)
- [Redis](https://redis.io) (Backend)
- [Weaviate](https://weaviate.io) (Backend)


## Dev requirements
You need to have `npm` and `docker-compose` commands available.

## Development Setup
This guide provides all necessary steps to have a fully functional local setup.

Clone this repository and access the repository root directory:

```
git clone git@github.com:gone-bike/gone.bike.git && \
cd gone.bike
```

---

### Frontend setup
Create development env file in `astro/` directory and install dependencies:

```
cd astro && \
cp .env.example .env && \
npm i
```

For pages that do not require db connection, this is enough to have the dev environment up and running, otherwise you need to have the backend setup.

Generate multilanguage pages:
```npx astro-i18next generate```

Start the local server with:
```npm run dev```

---

### Database setup

In repository root directory, use template env files:

```
touch .env && \
cp .worker.env.example .worker.env && \
cp .directus.env.example .directus.env && \
cp .astro.env.example .astro.env
```

Create local data and config folders:
```
mkdir -p .config .data/{directus,postgresql} && \
chmod 0777 .data/directus
```

Create local config folder and files:
```
touch .config/redis.conf
```

Startup `postgresql`, `redis` and `directus` services:

```
docker-compose up -d postgresql redis directus
```

Ensure server is running with:
```
docker-compose logs --tail 10 directus
```

that should show something smilar to this:
```
gone.bike.directus  | [12:35:20.420] INFO: Adding first admin user...
gone.bike.directus  | [12:35:20.483] INFO: Done
gone.bike.directus  | [12:35:23.050] INFO: Server started at http://0.0.0.0:8055
```

Reflect `DIRECTUS_TOKEN` in database (for website to access it):
```
export $(cat astro/.env | grep DIRECTUS_TOKEN | tr -d '"') && \
 docker-compose exec postgresql psql -U postgres -c "UPDATE directus_users SET token = '$DIRECTUS_TOKEN'"
```

Copy database schema inside directus container and apply it:

```
docker-compose cp database/directus-schema.yml directus:/directus/ && \
docker-compose exec directus npx directus schema apply -y directus-schema.yml
```

Verify successful schema install by accessing directus at [http://localhost:8055](http://localhost:8055) using default user and password (configurable in `.directus.env` file before first startup):

```
username: dev@gone.bike
password: dev
```


**NOTE**: It might be a directus bug, but in order to fully apply the schema and see it in the CMS, you need to perform a "Make collection invisible / Make collection visible" operation on any of the available collections. Operation can be applied twice in order to keep the state. After that, access the [/admin/settings/data-model/bike_brand](http://localhost:8055/admin/settings/data-model/bike_brand) address to enforce Directus metadata reload.

You should now be able to see a empty but schemed database.

---

### Database import

SQL database

```
wget https://gone.bike/exports/gone.bike.db-dump.latest.sql.gz && \
gunzip -c gone.bike.db-dump.latest.sql.gz | docker-compose exec -T postgresql psql -U postgres
```

Images

```
wget https://gone.bike/exports/gone.bike.images.latest.tgz  && \
tar zxvf gone.bike.images.latest.tgz --directory .data/directus/
```


### Image search setup
@TODO

CHANGELOG.md
# 2023-05-301 Setup guide, export procedures
- `filename_disk` to `filename_download` refactor
- Playing around with `rembg` in order to index pictures with removed bg (increased accuracy?)
- Worker componetization and event-driven approach
- Simple weaviate endpoint wrapper to display images stored within the db

# 2023-05-30 Setup guide, export procedures
- Fully-featured README file to help setting up a local dev
- Makefile commands to export data programmatically

# 2023-05-29 Report form
- Perfecting data acquisition via a user-friendly form

# 2023-05-28 Multi-locale experiments
- Set up to convert single-locale paths into multi-locale via `i18n-astro`

# 2023-05-27 OpenSourceSaturday, first public homepage release!
- Met `@fede03bill` who jumped in and helped on frontend side
- Spinned up production server
- Github Actions setup for auto-deploy on push to `main`
- QA on report form submissions
- Worker handling sumbissions and edgy cases
- Refined autotransation Flows

# 2023-05-26 Translations
- Scraping some similar website to gather an initial dataset
- Worker to accept input data coming from scraping

# 2023-05-24 Translations
- Draft of worker function to process submission

# 2023-05-23 Translations
- Started pushing translations
- Upload file endpoint

# 2023-05-22 Triggers, automation, dataflow
- Setup with Directus-to-Celery flow when editing reports
- Worker manages sync between databases
- Testing image search with simple file upload
- Report schema definition
- Tried and failed to use Jotform to avoid form implementation. Unable to fetch attachment via jotform API.

# 2023-05-21 - Git setup, staging setup
- Created https://github.com/gone-bike
- Pushed initial files
- photo controller wrapper using `filename_disk` too
- Base layout template

# 2023-05-20 - Data & software
- Evolved relational schema
- Added `i18n` schema
- Used ChatGPT to generate a list of bike brands, saved in db
- Setup Directus Flow to have automated translations within Directus
- AstroJS scaffolding and initial tests

# 2023-05-19 - Project setup
- Staging server
- Initial DB schema
- VPN setup (will be used to access the dabase)

# 2023-05-19 - Team buildup, initial config
- Job posting on Upwork, looking for collaborators, paid and voluntary.
- `docker-compose.yml` template to spin up a SQL server and a Directus instance.

# 2023-05-18 - Start
- Project start. Domain name picked: gone.bike (lost.bike, missing.bike taken, as usual)





worker/tasks.py
import sys, os, json, logging, time, requests, psycopg2, base64, dotmap, re

import utils

from slugify import slugify
from time import time as tm
from urllib.parse import urlparse

from celery import Celery
from celery.signals import celeryd_after_setup
from celery.signals import task_failure
from celery.exceptions import SoftTimeLimitExceeded

from weaviate.util import generate_uuid5

from geopy.geocoders import Nominatim
from geopy.extra.rate_limiter import RateLimiter

# take environment variables from .env.
#
from dotenv import load_dotenv
#load_dotenv()

geolocator = Nominatim(user_agent="gone.bike")
geocode = RateLimiter(geolocator.geocode, min_delay_seconds=10, error_wait_seconds=20)

app = Celery('tasks')
app.config_from_object({
    # 'task_default_queue': 'test',
    # 'task_default_exchange': 'test',
    # 'task_default_routing_key': 'test',
    'worker_prefetch_multiplier': os.environ["WORKER_PREFETCH_MULTIPLIER"],
    'broker_transport_options': {
        'visibility_timeout': 30, # BROKER_TRANSPORT_OPTTION_VISIBILITY_TIMEOUT
        'max_retries': 1
    }
})


@celeryd_after_setup.connect
def capture_worker_name(sender, instance, **kwargs):
    os.environ["WORKER_NAME"] = '{0}'.format(sender)



@app.task(bind=True, acks_late=False)
def test(self, *args, **kwargs):
    print(args)
    return True



@app.task(bind=True, acks_late=False)
def report_create(self, *args, **kwargs):
    # print(args)
    for i in args:
        item = utils.fetch_item(i)
        if item.get('status') == 'published':
            x = utils.index_directus_report_item_to_weaviate(item)
            print(x)

    return True

@app.task(bind=True, acks_late=False)
def report_update(self, *args, **kwargs):
    for i in args:
        item = utils.fetch_item(i)
        print(json.dumps(item,indent=2))

        # When an update occurs, it could be happening that an entry gets different associated photos
        for e in utils.fetch_unlinked_files():
            try:
                utils.remove_directus_file_by_id(e[0])
                utils.remove_weaviate_entry_by_id(e[0])
            except Exception as e:
                print(e)
                pass

        if item.get('status') == 'published':
            x = utils.index_directus_report_item_to_weaviate(item)
            print(x)
        else:
            utils.deindex_directus_report_item_from_weaviate(i)



    return True


@app.task(bind=True, acks_late=False)
def report_delete(self, *args):
    for i in args:
        utils.deindex_directus_report_item_from_weaviate(i)
    return True



### When a new report entry is submitted to the website, it enters a queue that is then processed by this function
@app.task(bind=True, acks_late=False)
def report_submit(self, *args, **kwargs):
    print(kwargs)


    # Bike brand / model normalization: if existing at db level, use it, otherwise create a new entry

    bike_brand = None
    bike_model = None

    if "bike_brand_id" in kwargs and kwargs.get('bike_brand_id') != "":
        bike_brand = int(kwargs.get('bike_brand_id'))

    elif "bike_brand" in kwargs and kwargs.get('bike_brand',"").strip() != "":
        url = f'{os.environ["WORKER_DIRECTUS_URI"]}/items/bike_brand?access_token={os.environ["WORKER_DIRECTUS_TOKEN"]}'
        bike_brand_slug = slugify(kwargs['bike_brand'])
        bike_brand = requests.get(f'{url}&filter[key][_eq]={ bike_brand_slug }')
        if bike_brand.status_code != 200:
            return {"status": "fail", "output": bike_brand.content }
        bike_brand = bike_brand.json()

        if len(bike_brand["data"]) == 0:
            bike_brand = requests.post(url, json={ "key": bike_brand_slug, "name": kwargs["bike_brand"] }).json()
            bike_brand = bike_brand["data"]["id"]

        else:
            bike_brand = bike_brand["data"][0]["id"]

    # if "bike_model_id" in kwargs and kwargs.get('bike_model_id') != "":
    #     bike_model = int(kwargs.get('bike_model_id'))

    if "bike_model" in kwargs and kwargs.get('bike_model','').strip() != "" and kwargs.get('bike_model') != None:
        url = f'{os.environ["WORKER_DIRECTUS_URI"]}/items/bike_brand_model?access_token={os.environ["WORKER_DIRECTUS_TOKEN"]}'

        bike_model_slug = slugify(kwargs['bike_model'])
        if bike_model_slug != "":
            bike_model = requests.get(f'{url}&filter[key][_eq]={ bike_model_slug }')
            if bike_model.status_code == 200:
                bike_model = bike_model.json()
                if len(bike_model["data"]) == 1:
                    bike_model = bike_model["data"][0]["id"]
                else:
                    bike_model = requests.post(url, json={ "bike_brand": bike_brand, "key": bike_model_slug, "name": kwargs["bike_model"] }).json()
                    bike_model = bike_model["data"]["id"]


    language= None
    if "language" in kwargs:
        url = f'{os.environ["WORKER_DIRECTUS_URI"]}/items/language?access_token={os.environ["WORKER_DIRECTUS_TOKEN"]}'

        language = requests.get(f'{url}&filter[locale_code][_eq]={ kwargs.get("language") }')
        if language.status_code == 200:
            language = language.json()
            if len(language["data"]) == 1:
                language = language["data"][0]["id"]


    # @TODO colors will be submitted in different languages, normalization function here is due
    try:
        colors = kwargs.get("colors","").replace('/',',').replace( 'and ', ',').split(',') if 'colors' in kwargs else []
        colors = list(map(lambda c : slugify(c) , colors))
    except Exception as e:
        colors = []


    approximate_value = 0
    try:
        approximate_value = int(kwargs.get("approximate_value", 0))
    except Exception as e:
        pass


    entry = {
        "bike_brand": bike_brand,
        "bike_brand_model": bike_model,

        "bike_details": kwargs.get("bike_details"),

        "approximate_value": approximate_value,
        "approximate_value_currency": kwargs.get("approximate_value_currency"),
        "colors": colors,
        "is_electric": bool(kwargs.get("is_electric")) if kwargs.get('is_electric') else None,
        "serial_number" : kwargs.get("serial_number"),
        "theft_status": kwargs.get('theft_status', 'stolen'),
        "theft_date": kwargs.get("theft_date"),
        "theft_timeframe": kwargs.get("theft_timeframe"),
        "theft_location_type": kwargs.get("theft_location_type"),
        "language" : language,
        "bike_type": kwargs.get('bike_type'),
        "lock_type": kwargs.get("lock_type"),
        "lock_anchor": kwargs.get("lock_anchor"),
        "location_address": kwargs.get("location_address"),

        "location_details": kwargs.get("location_details"),
        "description": kwargs.get("description"),
        "email": kwargs.get("email"),
        "ref_url": kwargs.get("ref_url"),
        "tags": kwargs.get('tags'),
        "submit_by": kwargs.get('submit_by')
    }

    print("Entry...")
    print(json.dumps(entry))

    # Location can be either explicit coordinates or an address that is then geocoded using Nominatim. Best effort approach
    if kwargs.get('location_coords'):
        entry['location'] = { "coordinates": [ kwargs.get("location_coords",{}).get('lng'), kwargs.get("location_coords",{}).get('lat') ], "type": "Point"}

    elif kwargs.get('location_address_raw'):
        location = geocode(kwargs.get('location_address_raw'))
        try:
            entry['location_address'] = location.address
            entry['location_details'] = location.raw
            entry['location_details']['_input'] = kwargs.get('location_address_raw')
            entry['location'] = { "coordinates": [ location.longitude, location.latitude ], "type": "Point"}
        except Exception as e:
            print(e)
            entry['location_details'] = {
                "_input": kwargs.get('location_address_raw')
            }



    # Files are sent as either full URLs or <upload,name> tuples. They are imported in Directus using directus API
    uploads = {}
    for pid in range(0,9):
        e = 'main_photo' if pid == 0 else f'photos_{pid}'

        if e in kwargs and kwargs[e] is not None and kwargs[e] != "":

            if type(kwargs[e]) is dict:
                photo_url = f'{os.environ["WORKER_PUBLIC_TMP_URL"]}/{kwargs[e]["upload"]}.{kwargs[e]["name"]}'
                photo_name = kwargs[e]["name"]
            if type(kwargs[e]) is str:
                photo_url = kwargs[e]
                photo_name = print(os.path.basename( urlparse(photo_url).path))

            print(f"Importing {photo_url}")


            # @TODO - env-based folder ids
            url = f'{os.environ["WORKER_DIRECTUS_URI"]}/files/import?access_token={os.environ["WORKER_DIRECTUS_TOKEN"]}'
            data = requests.post(url, json={
                "url": photo_url,
                "data": {
                    "folder": "fa2b1d0e-2e58-4897-af16-eab29f26117d" if e == 'main_photo' else "21f79529-85a1-4797-b4a1-7ddefdef654f",
                    "tags": [ photo_url ],
                    "description": photo_url,
                    "title": photo_name
                }
            })
            response = data.json()
            uploads[e] = response["data"]["id"]


    entry["main_photo"] = uploads.get('main_photo')

    entry["photos"] = { "create": [ ] }

    for x in uploads:
      if x != 'main_photo':
        entry["photos"]["create"].append({
          "report_id": "+",
          "directus_files_id": {
            "id": uploads.get(x)
          }
        })

    print(json.dumps(entry))

    url = f'{os.environ["WORKER_DIRECTUS_URI"]}/items/report?access_token={os.environ["WORKER_DIRECTUS_TOKEN"]}'
    data = requests.post(url, json=entry)

    # @TODO - some error control
    print(data.status_code)
    if data.status_code == 200:
        data = data.json()
        # print(data)
        print(f"http://localhost:8055/admin/content/report/{data['data']['id']}")
        if kwargs.get('email'):
            send = utils.send_activation_email(kwargs.get('language'), kwargs.get('email'), data['data']['id'], data['data']['activation_code'])
            print(send)

        return 'ok'



